### ISSUE TITLE
Wrong return type of overwritten property getter

### ISSUE THREAD

**Issue by Kaniee** [microsoft/pyright#4131]
**Describe the bug**
If a class overwrites just the property getter of the parent class, pyright assumes the return type of the parent getter instead of the sub class getter.

**To Reproduce**
```python
class PropBase:
    pass

class Base:
    @property
    def p(self) -> PropBase:
        return PropBase()
    
    @p.setter
    def p(self, p: PropBase) -> None:
        print("I don't want to write this twice.")

class PropSub(PropBase):
    def do_stuff(self) -> None:
        print("I did stuff.")

class Sub(Base):
    @Base.p.getter
    def p(self) -> PropSub:
        return PropSub()

p = Sub().p  # p is typed as "PropBase"
print(type(p))  # Output: <class '__main__.PropSub'>
p.do_stuff()  # Not expected: Cannot access member "do_stuff" for type "PropBase". Member "do_stuff" is unknown
```

**Expected behavior**
I expect to not get a error in the last line and that the variable `p` is typed as `PropSub` instead of `PropBase`.

**VS Code extension or command-line**
I'm using pylance in VSCode but pyright in the command-line returns the same error
```
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.275
/home/user/test.py
  /home/user/test.py:24:3 - error: Cannot access member "do_stuff" for type "PropBase"
    Member "do_stuff" is unknown (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.49sec
```

**Additional context**
I tried other alternatives to overwrite the property in a base class, but I either had to write the setter again or I lose the setter:
```python
class PropBase:
    pass

class Base:
    @property
    def p(self) -> PropBase:
        return PropBase()
    
    @p.setter
    def p(self, p: PropBase) -> None:
        print("I don't want to write this twice.")

class PropSub(PropBase):
    def do_stuff(self) -> None:
        print("I did stuff.")

class Sub1(Base):
    @Base.p.getter
    def p(self) -> PropSub:
        return PropSub()

class Sub2(Base):
    @property
    def p(self) -> PropSub:
        return PropSub()

    @p.setter
    def p(self, p: PropBase) -> None:
        print("I don't want to write this typo.")

class Sub3(Base):
    @property  # Expected: "p" incorrectly overrides property of same name in class "Base". Property method "fset" is missing in override
    def p(self) -> PropSub:
        return PropSub()
    
p = Base().p
print(type(p))  # <class '__main__.PropBase'>
p1 = Sub1().p  # p1 is wrongly typed as "PropBase"
print(type(p1))  # <class '__main__.PropSub'>
p2 = Sub2().p  # p2 is typed as "PropSub"
print(type(p2))  # <class '__main__.PropSub'>
p3 = Sub3().p  # p2 is typed as "PropSub"
print(type(p3))  # <class '__main__.PropSub'>

# p.do_stuff()  # Expected: Cannot access member "do_stuff" for type "PropBase". Member "do_stuff" is unknown 
p1.do_stuff()  # Not expected: Cannot access member "do_stuff" for type "PropBase". Member "do_stuff" is unknown
p2.do_stuff()
p3.do_stuff()

Base().p = PropBase() # Output: I don't want to write this twice.
Sub1().p = PropBase() # Output: I don't want to write this twice.
Sub2().p = PropBase() # Output: I don't want to write this typo.
# Sub3().p = PropBase()  # Expected: Cannot assign member "p" for type "Sub3". "PropBase" is incompatible with "property"
```

**Comment by erictraut** [comment]
Properties require significant special casing in a type checker. What you're doing here is highly abnormal, so I think it's reasonable for a type checker not to handle this case. Mypy generates a similar error in this case.

There will always be ways to subvert a static type checker if you try hard enough. If you want your code to work with pyright, you'll need to stick with a more standard design pattern. The workaround you've identified is a reasonable one.

**Comment by Kaniee** [comment]
My workarounds don't work either.
Sub2: Here I have to write duplicate code.
Sub3: Here pyright rightfully returns an error on the incorrect overwrite of the property because the setter is missing

How can I work around that?

**Comment by Kaniee** [comment]
I was now able to rewrite my code base, getting around this issue.
